File: app/src/main/java/org/zeroxlab/momodict/model/Store.java
package org.zeroxlab.momodict.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

public interface Store {

    int MAX_LENGTH = 1000;

    boolean addBook(@NonNull Book book);

    Book getBook(@NonNull String name);

    List<Book> getBooks();

    boolean removeBook(@NonNull String name);

    boolean addEntries(@NonNull List<Entry> entries);

    List<Entry> queryEntries(@Nullable String keyWord);

    List<Entry> queryEntries(@Nullable String keyWord, @NonNull String bookName);

    List<Entry> getEntries(@Nullable String keyWord);

    boolean upsertRecord(@NonNull Record record);

    boolean removeRecords(@NonNull String keyWord);

    List<Record> getRecords();

    boolean upsertCard(@NonNull Card card);

    boolean removeCards(@NonNull String keyWord);

    List<Card> getCards();
}

File: app/src/main/java/org/zeroxlab/momodict/model/Store.kt
package org.zeroxlab.momodict.model

interface Store {

    fun addBook(book: Book): Boolean

    fun getBook(name: String): Book

    fun getBooks(): List<Book>

    fun removeBook(bookName: String): Boolean

    fun addEntries(entries: List<Entry>): Boolean

    fun queryEntries(keyWord: String): MutableList<Entry>

    fun queryEntries(keyWord: String, bookName: String): MutableList<Entry>

    fun getEntries(keyWord: String): MutableList<Entry>

    fun getRecords(): MutableList<Record>

    fun upsertRecord(record: Record): Boolean

    fun removeRecords(keyWord: String): Boolean

    fun getCards(): MutableList<Card>

    fun upsertCard(card: Card): Boolean

    fun removeCards(keyWord: String): Boolean

    companion object {
        val MAX_LENGTH = 300
    }
}
